open_spiel/python/mfg/examples/mfg_Proximal_policy_optimization_pytorch.py

```diff
@@ -0,0 +1,448 @@
+
+
+def layer_init(layer, bias_const=0.0):
+    # used to initalize layers
+    nn.init.xavier_normal_(layer.weight)
+    nn.init.constant_(layer.bias, bias_const)
+    return layer
+
+
+class PPOpolicy(policy_std.Policy):
+    # required obeject to work with openspiel 
+    # used in updating the distribution using the policy 
+    # and in calculating the nash-convergance 
+
+    def __init__(self, game, agent, player_ids, device):
+        super().__init__(game, player_ids)
+        self.agent = agent
+        self.device = device
+
+    def action_probabilities(self, state, player_id=None):
+        # main method that is called to update the population states distribution
+        obs = torch.Tensor(state.observation_tensor()).to(self.device)
+        legal_actions = state.legal_actions()
+        logits = agent.actor(obs).detach().cpu()
+        legat_logits = np.array([logits[action] for action in legal_actions])
+        probs = np.exp(legat_logits -legat_logits.max())
+        probs /= probs.sum(axis=0)
+        
+        # returns a dictionary with actions as keys and their probabilities as values
+        return {action:probs[legal_actions.index(action)] for action in legal_actions}
+
+
+def rollout(env, iter_agent, eps_agent, num_epsiodes, steps, device):
+    # generates num_epsiodes rollouts
+    info_state = torch.zeros((steps,iter_agent.info_state_size), device=device)
+    actions = torch.zeros((steps,), device=device)
+    logprobs = torch.zeros((steps,), device=device)
+    rewards = torch.zeros((steps,), device=device)
+    dones = torch.zeros((steps,), device=device)
+    values = torch.zeros((steps,), device=device)
+    entropies = torch.zeros((steps,), device=device)
+    t_actions = torch.zeros((steps,), device=device)
+    t_logprobs = torch.zeros((steps,), device=device)
+
+    step = 0
+    for _ in range(num_epsiodes):
+        time_step = env.reset()
+        while not time_step.last():
+            obs = time_step.observations["info_state"][0]
+            obs = torch.Tensor(obs).to(device)
+            info_state[step] = obs
+            with torch.no_grad():
+                t_action, t_logprob, _, _ = iter_agent.get_action_and_value(obs)
+                action, logprob, entropy, value = eps_agent.get_action_and_value(obs)
+
+            time_step = env.step([action.item()])
+
+            # iteration policy data
+            t_logprobs[step] = t_logprob
+            t_actions[step] = t_action
+
+            # episode policy data
+            logprobs[step] = logprob
+            dones[step] = time_step.last()
+            entropies[step] = entropy
+            values[step] = value
+            actions[step] = action
+            rewards[step] = torch.Tensor(time_step.rewards).to(device)
+            step += 1
+
+    return info_state, actions, logprobs, rewards, dones, values, entropies,t_actions,t_logprobs 
+
+def cal_Adv(gamma, norm, rewards,values, dones):
```

Comment:
snake case: change to `cal_adv`

Reply:
Style of functions changed to sanke_case